aat/benchmarkrpc_test.go

```diff
@@ -0,0 +1,143 @@
+package aat
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"testing"
+
+	"github.com/gammazero/nexus/client"
+	"github.com/gammazero/nexus/wamp"
+)
+
+func BenchmarkRpcIntegerList(b *testing.B) {
+
+	args := wamp.List{make([]int, 128)}
+
+	benchmarkRpc(b, sum, args, func(result *wamp.Result) {
+		v, ok := wamp.AsInt64(result.Arguments[0])
+		if !ok {
+			panic("Can not typecast result to int64")
+		}
+		if v != 0 {
+			panic(fmt.Sprintf("Wrong result!: %d", v))
+		}
+
+	})
+
+}
+
+func BenchmarkRpcShortString(b *testing.B) {
+
+	shortString := randomString(128)
+	benchmarkRpc(b, identify, wamp.List{shortString}, func(result *wamp.Result) {
+		v, ok := wamp.AsString(result.Arguments[0])
+		if !ok {
+			panic("Can not typecast result to int64")
+		}
+		if v != shortString {
+			panic(fmt.Sprintf("Wrong result!: %v", v))
+		}
+
+	})
+
+}
+
+func BenchmarkRpcLargeString(b *testing.B) {
+
+	largeString := randomString(4096)
+	benchmarkRpc(b, identify, wamp.List{largeString}, func(result *wamp.Result) {
+		v, ok := wamp.AsString(result.Arguments[0])
+		if !ok {
+			panic("Can not typecast result to int64")
+		}
+		if v != largeString {
+			panic(fmt.Sprintf("Wrong result!: %v", v))
+		}
+
+	})
+
+}
+
+func BenchmarkRpcDict(b *testing.B) {
+
+	dict := wamp.Dict{}
+	for i := 0; i < 8; i++ {
+		dict[randomString(8)] = randomString(8)
+	}
+	args := wamp.List{dict}
+
+	benchmarkRpc(b, identify, args, func(result *wamp.Result) {
+		v, ok := wamp.AsDict(result.Arguments[0])
+		if !ok {
+			panic("Can not typecast result to int64")
+		}
+		if len(v) != 8 {
+			panic(fmt.Sprintf("Wrong result!: %d", len(v)))
+		}
+
+	})
+}
+
+func benchmarkRpc(b *testing.B, action client.InvocationHandler, callArgs wamp.List, verify func(*wamp.Result)) {
+
+	server, err := connectClient()
+	if err != nil {
+		panic("Failed to connect client: " + err.Error())
+	}
+
+	if err = server.Register("action", action, nil); err != nil {
+		panic("Failed to register procedure: " + err.Error())
+	}
+
+	client, err := connectClient()
+	if err != nil {
+		panic("Failed to connect client: " + err.Error())
+	}
+
+	ctx := context.Background()
+
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+
+		result, err := client.Call(ctx, "action", nil, callArgs, nil, "")
+		if err != nil {
+			panic(err)
+		}
+		if i < 2 {
+			verify(result)
```

Comment:
Even though the number of calls to `verify()` is limited, you may want to consider calling it outside of the benchmark loop, after calling `b.StopTimer()`

Reply:
A note about benchmarking nexus in general: The nexus router is designed to be highly concurrent.  It tries to handle as much as possible on concurrent goroutines, except for the aspects or routing that need access to shared resources.

What this means is that a benchmark for traffic between any two clients will not be better than a  totally serialized/single-threaded router.  The performance advantage of nexus should be seen as the number of clients increases.  A benchmark that ramps up the number of clients would also be good.